elide-datastore/elide-datastore-aggregation/src/main/java/com/yahoo/elide/datastores/aggregation/queryengines/sql/metric/SQLBasicMetricFunction.java

```diff
@@ -0,0 +1,120 @@
+/*
+ * Copyright 2019, Yahoo Inc.
+ * Licensed under the Apache License, Version 2.0
+ * See LICENSE file in project root for terms.
+ */
+package com.yahoo.elide.datastores.aggregation.queryengines.sql.metric;
+
+import com.yahoo.elide.core.exceptions.InternalServerErrorException;
+import com.yahoo.elide.core.exceptions.InvalidOperationException;
+import com.yahoo.elide.datastores.aggregation.metadata.metric.AggregatedField;
+import com.yahoo.elide.datastores.aggregation.metadata.metric.BasicMetricFunction;
+import com.yahoo.elide.datastores.aggregation.metadata.metric.MetricFunctionInvocation;
+import com.yahoo.elide.datastores.aggregation.metadata.models.FunctionArgument;
+import com.yahoo.elide.datastores.aggregation.metadata.models.Metric;
+import com.yahoo.elide.datastores.aggregation.query.DimensionProjection;
+import com.yahoo.elide.datastores.aggregation.query.TimeDimensionProjection;
+import com.yahoo.elide.datastores.aggregation.queryengines.sql.query.SQLQueryTemplate;
+import com.yahoo.elide.request.Argument;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+/**
+ * SQL extension of {@link BasicMetricFunction} which would be invoked as sql and can construct sql templates.
+ */
+public abstract class SQLBasicMetricFunction extends BasicMetricFunction {
+    public SQLBasicMetricFunction(String name, String longName, String description, Set<FunctionArgument> arguments) {
```

Comment:
I think we should collapse the entire Metric hierarchy into fewer concepts.  Let's just have MetricFunctions and SQLMetricFunctions.

Reply:
👍